src/pages/node-express/lesson13-a1.mdx

```diff
@@ -0,0 +1,526 @@
+  res.render("secretWord", { secretWord: req.session.secretWord });
+});
+
+router.post("/", (req, res) => {
+  if (req.body.secretWord.toUpperCase()[0] == "P") {
+    req.flash("error", "That word won't work!");
+    req.flash("error", "You can't use words that start with p.");
+  } else {
+    req.session.secretWord = req.body.secretWord;
+    req.flash("info", "The secret word was changed.");
+  }
+
+  res.redirect("/secretWord");
+});
+
+module.exports = router;
+```
+
+We could further refactor this by moving the code for handling the routes (`(req, res) => {...}`) a controller file, like we've
+done for the session routes, but we'll leave it like this for now.
+
+You'll want to update `views/secretWord.ejs` to display the `req.flash`.
+You can follow a similar pattern as what's used in `views/partials/header.ejs`. You'll also want to add a button for
+logging off on that page (create a `<form>` element that makes a `POST` call to `/sessions/logoff`).
+
+Then replace the `app.get` and `app.post` statements for the `"/secretWord"` routes in `app.js` with these lines.
+
+```js
+const secretWordRouter = require("./routes/secretWord");
+app.use("/secretWord", secretWordRouter);
+```
+
+Then try out the secretWord page to make sure it still works. Turning on protection is simple.
+You add the authentication middleware to the route above as follows:
+
+```js
+const auth = require("./middleware/auth");
+app.use("/secretWord", auth, secretWordRouter);
+```
+
+That causes the authentication middleware to run before the `secretWordRouter`, and it redirects
+if any requests are made for those routes before logon. Try it out: login and verify that
+you can see and change the secretWord. Then logoff and try to go to the `"/secretWord"` URL.
+
+## Fixing the Security
+
+Passport is using the session cookie to determine if the user is logged in. This creates a security vulnerability called "cross site request forgery" (CSRF). We will demonstrate this.
+
+To see this, clone **[this repository](https://github.com/Code-the-Dream-School/csrf-attack)** into a separate directory, outside of the current `jobs-ejs` folder. Then, within the directory you cloned, install packages with `npm install` and run the app with `node app`. This will start another express application listening on port **4000** of your local machine. This is the attacking code. It could be running anywhere on the Internet -- that has nothing to do with the attack.
+
+You should have two browser tabs open, one for localhost:3000, and one for localhost:4000. The one at localhost:4000 just shows a button that says Click Me! **Don't click it yet**. Use the `jobs-ejs`
+application in the 3000 tab to set the secret string to some value. Then log off.
+Then click the button in the 4000 tab. Then log back on in the 3000 tab and check the value of the secret string. So far so good -- it still has the value you set.
+
+[//]: # "# TODO: Verify this exercise because logging off will delete from session"
```

Comment:
@jrmcgarvey I'm not sure if I missed something when going through this steps but I wasn't able to get the following steps to work:  
\> Use the \`jobs-ejs\`  
application in the 3000 tab to set the secret string to some value. Then log off.  
Then click the button in the 4000 tab. Then log back on in the 3000 tab and check the value of the secret string. So far so good -- it still has the value you set.

Once I log off, when I log back in, the word is reset to `syzygy`. I think it's because we're storing the word in the session, which means it persists across server restarts; but it still gets deleted when I log off and the session is destroyed. And when I log in again, that's a brand new fresh session.

Should we reword that part or is there a simple approach we could take for storing the user's word in db so that it persists across sessions?

Reply:
I need to change this -- it's incorrect.  It should say, close the browser tab with the application and then open it again.  I'll fix that.